ntoskrnl/mm/ARM3/pool.c

```diff
@@ -1580,6 +1580,10 @@ MmAllocateMappingAddress(
     PMMPTE PointerPte;
     MMPTE TempPte;
 
+    /* Fast exit if PoolTag is NULL */
+    if (!PoolTag)
```

Comment:
Is this behaviour documented? And tested?

Reply:
This is tested in the kmtest:MmReservedMapping by the following code:
```
    /* PoolTag = 0 */
    Mapping = MmAllocateMappingAddress(1, 0);
    ok(Mapping == NULL, "MmAllocateMappingAddress failed\n");
    if (Mapping != NULL)
    {
        MmFreeMappingAddress(Mapping, 0);
    }
    ```

Reply:
yeah but the link you showed is for a completely different function.

Reply:
OK. Please allow me to correct my semantics and revised the "fairly well" in my statement to "somewhat" and apologize for such an egregious mistake.

The actual function being tested here is "MmAllocateMappingAddress" which is documented here:

https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-mmallocatemappingaddress

Under the ```[in] PoolTag`` we see the following:

Specifies a four-character tag used to identify the buffer. Use a distinct PoolTag tag for each allocation code path. For a description of pool tags, see [ExAllocatePoolWithTag](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-exallocatepoolwithtag).

Now if we follow the link at the end of the previous sentence, we get to the link that I referenced above.

Under the heading ```Return value``` we have the following:

ExAllocatePoolWithTag returns NULL if there is insufficient memory in the free pool to satisfy the request. Otherwise, the routine returns a pointer to the allocated memory.

So you are correct that we do not have a directly derived relationship here, but only one that is indirect.
But the way that I read this points to the fact that the return value of some "Allocate" functions can return NULL.

And if you are not willing to accept this logic, then look at the test. It does stand on its own.
Look the these results from the "Test WHS" bot:
https://reactos.org/testman/compare.php?ids=98525,98544,98557,98619

You can see that for the ```kmtest:MmReservedMapping``` there are no failed tests.

Thanks for your review and comments.